File: homework_two/src/main/java/com/abra/alexhomework/MainActivity.java
package com.abra.alexhomework;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
/**
 * Главный класс, отвечает за отправку сгенерированных чисел
 * в другую активность.
 * В классе есть два решения: одно с использованием Intent,
 * второе с использованием паттерна Observer
 * Для удобного просмотра текст выводится в TextView, который находится в ScrollView
 * */
public class MainActivity extends AppCompatActivity implements Observer {
    /*
    * Объекты data и manager нужны для решения с паттерном Observer
    * раскомментируйте их для демонстрации решения
    */
//    private Data result;
//    private final DataManager manager = DataManager.getInstance();
    private static final int REQUEST_CODE = 3;
    private TextView text;
    private final NumberGenerator generator = new NumberGenerator();
    private ArrayList<Integer> numbers;
    @SuppressLint({"SetTextI18n", "DefaultLocale"})
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        text = findViewById(R.id.text1);
        Button bt = findViewById(R.id.button2);
        bt.setOnClickListener((view -> {
            numbers = generator.generateNumbers();
            Intent intent = new Intent(this,CountActivity.class);
            intent.putExtra("numbers",numbers);
            startActivityForResult(intent,REQUEST_CODE);
// код написаный ниже нужен для решения с паттерном Observer, раскомментируйте для демонстрации
//            numbers = generator.generateNumbers();
//            manager.addObserver(this);
//            Intent intent = new Intent(this,CountActivity.class);
//            manager.notifyAllObservers(new Data(numbers,0,0,0));
//            startActivityForResult(intent,REQUEST_CODE);
        }));

    }
    @SuppressLint({"SetTextI18n", "DefaultLocale"})
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if(data!=null){
            int sumResult;
            double halfResult, arithmeticMeanResult;
            sumResult = data.getIntExtra("sumResult",0);
            arithmeticMeanResult = data.getDoubleExtra("arithmeticMeanResult",0);
            halfResult = data.getDoubleExtra("halfResult",0);
            text.setText(getString(R.string.generated_numbers)+"\n"+ numbers.toString() +"\n"+
                    getString(R.string.sum)+" "+sumResult+"\n"+
                    getString(R.string.arithmetic_mean)+" "+String.format("%.2f",arithmeticMeanResult)+"\n"+
                            getString(R.string.devision)+" "+String.format("%.2f",halfResult));

// код написаный ниже нужен для решения с паттерном Observer, раскомментируйте для демонстрации
//            text.setText(getString(R.string.generated_numbers)+"\n"+ numbers.toString() +"\n"+
//                    getString(R.string.sum)+" "+result.getSumResult()+"\n"+
//                    getString(R.string.arithmetic_mean)+" "+String.format("%.2f",result.getArithmeticMeanResult())+"\n"+
//                    getString(R.string.devision)+" "+String.format("%.2f",result.getHalfResult()));

        }
    }
    @Override
    public void update(Data data) {
        // код написаный ниже нужен для решения с паттерном Observer, раскомментируйте для демонстрации
//        result = data;
    }
}

File: homework_5/src/main/java/com/abra/homework_5/MainActivity.kt
package com.abra.homework_5

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.widget.SearchView
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.io.File
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity(), CarInfoAdapter.OnEditIconClickListener, CarInfoAdapter.OnCarInfoClickListener {
    private lateinit var fab: FloatingActionButton
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: CarInfoAdapter
    private lateinit var noCarsAddedText: TextView
    private lateinit var database: DataBaseCarInfo
    private lateinit var carInfoDAO: CarInfoDAO
    private lateinit var searchView: SearchView
    private val ADD_ACTIVITY_CODE = 1
    private val EDDIT_ACTIVITY_CODE = 2
    private val WORK_LIST_ACTIVITY = 10
    private val RESULT_CODE_BUTTON_BACK = 5
    private var searchWasUsed: Boolean = false
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        fab = findViewById(R.id.fabAddCar)
        recyclerView = findViewById(R.id.recyclerViewCars)
        noCarsAddedText = findViewById(R.id.tvNoCarsAdded)
        database = DataBaseCarInfo.getDataBase(this)
        carInfoDAO = database.getCarInfoDAO()
        adapter = if (carInfoDAO.getAll().isNotEmpty()) {
            CarInfoAdapter(this, carInfoDAO.getAll()).also { it.sortByCarName(carInfoDAO.getAll()) }
        } else {
            CarInfoAdapter(this)
        }
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        setFabListener()
        adapter.setOnEditIconClickListener(this)
        adapter.setOnCarInfoClickListener(this)
        writeLogToFile()
        if (adapter.itemCount != 0) noCarsAddedText.visibility = View.INVISIBLE
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.main_menu, menu)
        searchView = menu?.findItem(R.id.search)?.actionView as SearchView
        searchView.imeOptions = EditorInfo.IME_ACTION_DONE
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(p0: String?) = false

            override fun onQueryTextChange(p0: String?): Boolean {
                searchWasUsed = true
                adapter.filter.filter(p0)
                return false
            }
        })
        return true
    }

    private fun setFabListener() {
        fab.setOnClickListener {
            val intent = Intent(this, AddCarActivity::class.java)
            startActivityForResult(intent, ADD_ACTIVITY_CODE)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != RESULT_CODE_BUTTON_BACK) {
            adapter.sortByCarName(carInfoDAO.getAll())
            adapter.sortByCarName(carInfoDAO.getAll())
            if (searchWasUsed) {
                searchView.onActionViewCollapsed()
                searchWasUsed = false
            }
        }
        if (adapter.itemCount != 0) noCarsAddedText.visibility = View.INVISIBLE
    }

    override fun onEditIconClick(carInfo: CarInfo) {
        val intent = Intent(this, EditCarInfoActivity::class.java)
        intent.putExtra("carInfo", carInfo)
        startActivityForResult(intent, EDDIT_ACTIVITY_CODE)
    }

    override fun onCarInfoClick(carInfo: CarInfo) {
        val intent = Intent(this, WorkListActivity::class.java)
        intent.putExtra("carInfo", carInfo)
        startActivityForResult(intent, WORK_LIST_ACTIVITY)
    }

    private fun writeLogToFile() {
        val logList = StringBuilder()
        val file = File(filesDir, "Logs.txt")
        if (!file.exists()) {
            file.createNewFile()
        }
        logList.append("${file.readText()} \n")
        val simpleDateFormat = SimpleDateFormat.getDateTimeInstance()
        val date = simpleDateFormat.format(Date())
        logList.append(date)
        file.writeText(logList.toString())
    }
}